src/profile-logic/symbol-store.js

```diff
@@ -361,82 +362,121 @@ export class SymbolStore {
     // Process the results from the symbolication API request, as they arrive.
     // For unsuccessful requests, fall back to _getSymbolsFromBrowser.
     await Promise.all(
-      chunkResponsePromises.map(async ([requests, chunkResponsePromise]) => {
-        try {
-          // Await the response for this chunk.
-          const responses: LibSymbolicationResponse[] =
-            await chunkResponsePromise;
-
-          const requestsWithErrors = [];
-          for (const response of responses) {
-            if (response.type === 'SUCCESS') {
-              // We've found symbols for this library! Option 2 was successful.
-              const { lib, results } = response;
-              successCb(lib, results);
-            } else {
-              // Collect unsuccessful requests in requestsWithErrors.
-              // We keep the error around so that we can report it if all avenues fail.
-              const { request, error } = response;
-              requestsWithErrors.push({ request, error });
-            }
+      symbolicationApiRequestsAndResponsesPerChunk.map(
+        async ([option2Requests, option2ResponsePromise]) => {
+          // Store any errors encountered along the way in this map.
+          // We will report them if all avenues fail.
+          const errorMap: Map<LibSymbolicationRequest, Error[]> = new Map(
+            requests.map((r) => [r, []])
+          );
+
+          // Process the results of option 2: The response from the Mozilla symbolication APi.
+          const option3Requests =
+            await this._processSuccessfulResponsesAndReturnRemaining(
+              option2Requests,
+              option2ResponsePromise,
+              errorMap,
+              successCb
+            );
+
+          if (option3Requests.length === 0) {
+            // Done!
+            return;
           }
 
-          if (requestsWithErrors.length > 0) {
-            // Fall back to getting symbols from the browser for these.
-            await this._getSymbolsFromBrowser(
-              requestsWithErrors,
-              demangleCallback,
-              successCb,
-              errorCb
+          // Some libraries are still remaining, try option 3 for them.
+          // Option 3 is symbolProvider.requestSymbolsFromBrowser.
+          const option3ResponsePromise =
+            this._symbolProvider.requestSymbolsFromBrowser(option3Requests);
```

Comment:
The name of the functions `requestSymbolsFromBrowser` vs `requestSymbolTableFromBrowser` are very close, do you have an idea to make them more descriptive? Maybe `requestSymbolsFromWebChannel` vs `requestSymbolTableFromFrameScript`? But for Android we're not using the framescript, are we?